howso/direct/core.py

```diff
@@ -3142,6 +3159,55 @@ def _execute_sized(self, label: str, payload: Any) -> Tuple[Any, int, int]:
             return None, len(json_payload), 0
         return self._deserialize(result), len(json_payload), len(result)
 
+    def _verify_trainee_header(self, trainee_filepath: Path):
+        """
+        Verifies the trainee header for compatibility with Amalgam binary.
+
+        Parameters
+        ----------
+        trainee_filepath : str
+            Path to trainee.
+        """
+
+        # Parse Amalgam version, removing suffix:
+        amalgam_version = self.amlg.get_version_string().decode("utf-8")
+        amalgam_major, amalgam_minor, amalgam_patch = map(int, amalgam_version.split('-')[0].split('.'))
+
+        # If the file doesn't exist, we let the normal code path deal with that. This function
+        # is specifically for checking the header of an existing file.
+        if trainee_filepath.exists():
+            with open(trainee_filepath, "rb") as f:
+
+                # Check magic number to make sure we are looking at a CAML file:
+                magic_number = f.read(4)
+                if magic_number.decode('utf-8') != "caml":
+                    raise HowsoError('Trainee has a malformed header')
+
+                # Read trainee version as big endian ints:
+                trainee_major = int(struct.unpack('>i', f.read(4)))
+                trainee_minor = int(struct.unpack('>i', f.read(4)))
+                trainee_patch = int(struct.unpack('>i', f.read(4)))
+                trainee_version = f"{trainee_major}.{trainee_minor}.{trainee_patch}"
+
+                # Check trainee is not newer than amalgam:
+                if (
+                    (trainee_major > amalgam_major) or
+                    (trainee_major == amalgam_major and trainee_minor > amalgam_minor) or
+                    (trainee_major == amalgam_major and trainee_minor == amalgam_minor and
+                        trainee_patch > amalgam_patch)
+                ):
+                    raise HowsoError(
```

Comment:
I'm not crazy about raising here. Can we not warn?

Reply:
Moved to warnings, not exceptions.